test/lib/tpunit++.cpp

```diff
@@ -389,28 +391,43 @@ void tpunit::TestFixture::tpunit_detail_do_methods(tpunit::TestFixture::method*
 
 void tpunit::TestFixture::tpunit_detail_do_tests(TestFixture* f) {
     method* t = f->_tests;
-    recursive_mutex& m = *(f->_mutex);
+    list<thread> testThreads;
     while(t) {
-       int _prev_assertions = f->_stats._assertions;
-       int _prev_exceptions = f->_stats._exceptions;
-       f->testOutputBuffer = "";
-       tpunit_detail_do_methods(f->_befores);
-       tpunit_detail_do_method(t);
-       tpunit_detail_do_methods(f->_afters);
-       if(_prev_assertions == f->_stats._assertions && _prev_exceptions == f->_stats._exceptions) {
-          lock_guard<recursive_mutex> lock(m);
-          printf("\xE2\x9C\x85 %s\n", t->_name);
-          tpunit_detail_stats()._passes++;
-       } else {
-          lock_guard<recursive_mutex> lock(m);
-
-          // Dump the test buffer if the test included any log lines.
-          f->printTestBuffer();
-          printf("\xE2\x9D\x8C !FAILED! \xE2\x9D\x8C %s\n", t->_name);
-          tpunit_detail_stats()._failures++;
-          tpunit_detail_stats()._failureNames.emplace(t->_name);
-       }
-       t = t->_next;
+        testThreads.push_back(thread([t, f]() {
+            recursive_mutex& m = *(f->_mutex);
+            f->_stats._assertions = 0;
+            f->_stats._exceptions = 0;
+            f->testOutputBuffer = "";
+            tpunit_detail_do_methods(f->_befores);
+            tpunit_detail_do_method(t);
+            tpunit_detail_do_methods(f->_afters);
+
+            // No new assertions or exceptions. This not currently synchronized correctly. They can cause tests that
+            // passed to appear failed when another test failed while this test was running. They cannot cause failed
+            // tests to appear to have passed.
+            if(!f->_stats._assertions && !f->_stats._exceptions) {
+                lock_guard<recursive_mutex> lock(m);
+                printf("\xE2\x9C\x85 %s\n", t->_name);
+                tpunit_detail_stats()._passes++;
+            } else {
+                lock_guard<recursive_mutex> lock(m);
```

Comment:
since the first line in both of these is to lock the mutex should we just move that above the `if !f->_stats._assertions && !f->_stats._exceptions)`? er hmm no i guess not because with it inside of the if/else blocks the lock_guard gets destroyed once we hit the end of the `{}` which then frees up the mutex, right?

Reply:
We probably could move it out but then like you say the scope of the lock is longer.

Reply:
👍 coolio I think leaving as is works well.